### PR TITLE
*: put gRPC unknown error into the header.

### LINKED ISSUE #5161: The error returned by PD gRPC service is not expected
## Bug Report

See https://github.com/tikv/pd/blob/d263b8586123387d219bdccf9a1733632ab0d9c2/server/grpc_service.go#L470

We should put the error into the header instead of directly returning a gRPC error.

### What version of PD are you using (`pd-server -V`)?
master

### PR DESCRIPTION
**Pull request by HuSharp**
Signed-off-by: husharp <jinhao.hu@pingcap.com>

### What problem does this PR solve?
https://github.com/tikv/pd/issues/5161

Issue Number: Close #5161 

### What is changed and how does it work?

```commit-message
put the unknown error into the header instead of directly returning a gRPC error.
```

### Check List

Tests
- Unit test

### Release note

```release-note
Put the unknown error into the header instead of directly returning a gRPC error.
And modify related uint tests and code.
```

### CODE REVIEW

**Review by Connor1996 — APPROVED** [review]
LGTM

File: server/grpc_service.go [hunk 1]
```diff
@@ -98,7 +98,12 @@ func (s *GrpcServer) GetMembers(context.Context, *pdpb.GetMembersRequest) (*pdpb
 	// at startup and needs to get the cluster ID with the first request (i.e. GetMembers).
 	members, err := s.Server.GetMembers()
 	if err != nil {
-		return nil, status.Errorf(codes.Unknown, err.Error())
+		return &pdpb.GetMembersResponse{
+			Header: s.errorHeader(&pdpb.Error{
+				Type:    pdpb.ErrorType_UNKNOWN,
+				Message: err.Error(),
+			}),
```
Comment (rleungx) [inline server/grpc_service.go:105]: How about:
```
func (s *GrpcServer) unknownErrorHeader(message string) *pdpb.ResponseHeader {
	return s.errorHeader(&pdpb.Error{
		Type:    pdpb.ErrorType_UNKNOWN,
		Message: message,
	})
}
```

File: server/util.go [hunk 2]
```diff
@@ -157,3 +157,10 @@ func checkBootstrapRequest(clusterID uint64, req *pdpb.BootstrapRequest) error {
 
 	return nil
 }
+
+func (s *GrpcServer) wrapErrorToHeader(errorType pdpb.ErrorType, message string) *pdpb.ResponseHeader {
```
Comment (JmPotato) [inline server/util.go:161]: I think we could just place this function in `grpc_service.go`.

File: server/grpc_service.go [hunk 3]
```diff
@@ -1487,12 +1520,17 @@ func (s *GrpcServer) SyncMaxTS(_ context.Context, request *pdpb.SyncMaxTSRequest
 	tsoAllocatorManager := s.GetTSOAllocatorManager()
 	// There is no dc-location found in this server, return err.
 	if tsoAllocatorManager.GetClusterDCLocationsNumber() == 0 {
-		return nil, status.Errorf(codes.Unknown, "empty cluster dc-location found, checker may not work properly")
+		return &pdpb.SyncMaxTSResponse{
```
Comment (JmPotato) [inline server/grpc_service.go:1523]: This gRPC method is used by the PD internally. I think you should also modify the caller to check the header error.

File: server/grpc_service.go [hunk 4]
```diff
@@ -1649,7 +1697,10 @@ func (s *GrpcServer) GetDCLocationInfo(ctx context.Context, request *pdpb.GetDCL
 	info, ok := am.GetDCLocationInfo(request.GetDcLocation())
 	if !ok {
 		am.ClusterDCLocationChecker()
-		return nil, status.Errorf(codes.Unknown, "dc-location %s is not found", request.GetDcLocation())
+		return &pdpb.GetDCLocationInfoResponse{
```
Comment (JmPotato) [inline server/grpc_service.go:1700]: Ditto.

**Review by JmPotato — COMMENTED** [review]
`server/grpc_service.go 100644 → 100755`.

Please change the file access permission back.

File: client/base_client.go [hunk 5]
```diff
@@ -328,9 +328,10 @@ func (c *baseClient) getMembers(ctx context.Context, url string, timeout time.Du
 	if err != nil {
 		return nil, err
 	}
-	members, err := pdpb.NewPDClient(cc).GetMembers(ctx, &pdpb.GetMembersRequest{})
-	if err != nil {
-		attachErr := errors.Errorf("error:%s target:%s status:%s", err, cc.Target(), cc.GetState().String())
+	members, _ := pdpb.NewPDClient(cc).GetMembers(ctx, &pdpb.GetMembersRequest{})
```
Comment (JmPotato) [inline client/base_client.go:331]: I think we can not just ignore the error here. The error inside header and the gRPC error should be returned both.

File: client/client.go [hunk 6]
```diff
@@ -583,12 +583,12 @@ func (c *client) GetAllMembers(ctx context.Context) ([]*pdpb.Member, error) {
 	ctx, cancel := context.WithTimeout(ctx, c.option.timeout)
 	req := &pdpb.GetMembersRequest{Header: c.requestHeader()}
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
-	resp, err := c.getClient().GetMembers(ctx, req)
+	resp, _ := c.getClient().GetMembers(ctx, req)
```
Comment (JmPotato) [inline client/client.go:586]: Ditto.

File: server/tso/allocator_manager.go [hunk 7]
```diff
@@ -1074,6 +1075,9 @@ func (am *AllocatorManager) getDCLocationInfoFromLeader(ctx context.Context, dcL
 	if err != nil {
 		return false, &pdpb.GetDCLocationInfoResponse{}, err
 	}
+	if resp.Header.GetError() != nil {
+		return false, &pdpb.GetDCLocationInfoResponse{}, errors.Errorf("get the dc-location info failed: %s", resp.Header.GetError().String())
```
Comment (JmPotato) [inline server/tso/allocator_manager.go:1079]:
```suggestion
		return false, &pdpb.GetDCLocationInfoResponse{}, errors.Errorf("get the dc-location info from leader failed: %s", resp.Header.GetError().String())
```

File: server/tso/global_allocator.go [hunk 8]
```diff
@@ -347,6 +348,11 @@ func (gta *GlobalTSOAllocator) SyncMaxTS(
 					log.Error("sync max ts rpc failed, got an error", zap.String("local-allocator-leader-url", leaderConn.Target()), errs.ZapError(err))
 					return
 				}
+				if syncMaxTSResp.rpcRes != nil && syncMaxTSResp.rpcRes.Header.GetError() != nil {
+					log.Error("sync max ts rpc failed, got an error", zap.String("local-allocator-leader-url", leaderConn.Target()),
+						errs.ZapError(errors.Errorf("get the dc-location info failed: %s", syncMaxTSResp.rpcRes.Header.GetError().String())))
```
Comment (JmPotato) [inline server/tso/global_allocator.go:353]: Ditto.

File: client/base_client.go (hunk 5 above)
Comment (JmPotato) [inline client/base_client.go:331]: I think we can not just ignore the gRPC error here. It should be returned also.

File: client/client.go (hunk 6 above)
Comment (JmPotato) [inline client/client.go:586]: Ditto.

File: server/api/member.go [hunk 9]
```diff
@@ -63,9 +63,9 @@ func (h *memberHandler) GetMembers(w http.ResponseWriter, r *http.Request) {
 func getMembers(svr *server.Server) (*pdpb.GetMembersResponse, error) {
 	req := &pdpb.GetMembersRequest{Header: &pdpb.RequestHeader{ClusterId: svr.ClusterID()}}
 	grpcServer := &server.GrpcServer{Server: svr}
-	members, err := grpcServer.GetMembers(context.Background(), req)
-	if err != nil {
-		return nil, errors.WithStack(err)
+	members, _ := grpcServer.GetMembers(context.Background(), req)
```
Comment (JmPotato) [inline server/api/member.go:66]: Ditto.

File: server/tso/allocator_manager.go (hunk 7 above)
Comment (JmPotato) [inline server/tso/allocator_manager.go:1079]:
```suggestion
		return false, &pdpb.GetDCLocationInfoResponse{}, errors.Errorf("get the dc-location info from leader failed: %s", resp.Header.GetError().String())
```

File: server/tso/global_allocator.go (hunk 8 above)
Comment (JmPotato) [inline server/tso/global_allocator.go:353]:
```suggestion
						errs.ZapError(errors.Errorf("%s", syncMaxTSResp.rpcRes.Header.GetError().String())))
```

File: tests/cluster.go [hunk 10]
```diff
@@ -260,9 +260,9 @@ func (s *TestServer) GetEtcdLeader() (string, error) {
 	s.RLock()
 	defer s.RUnlock()
 	req := &pdpb.GetMembersRequest{Header: &pdpb.RequestHeader{ClusterId: s.server.ClusterID()}}
-	members, err := s.grpcServer.GetMembers(context.TODO(), req)
-	if err != nil {
-		return "", errors.WithStack(err)
+	members, _ := s.grpcServer.GetMembers(context.TODO(), req)
```
Comment (JmPotato) [inline tests/cluster.go:263]: Same as the previous question.

File: client/base_client.go (hunk 5 above)
Comment (HuSharp) [reply client/base_client.go:331]: > I think we can not just ignore the error here. The error inside header and the gRPC error should be returned both.

I do this because the function GetMembers does not return an error.

File: tools/pd-simulator/simulator/client.go [hunk 11]
```diff
@@ -263,13 +269,16 @@ func (c *client) Bootstrap(ctx context.Context, store *metapb.Store, region *met
...
 		Region: proto.Clone(region).(*metapb.Region),
 	})
-	if err != nil {
-		return err
+	if err != nil || res.GetHeader().GetError() != nil {
+		if err != nil {
+			return err
+		}
+		return errors.Errorf("bootstrap failed: %s", resp.Header.GetError().String())
 	}
```
Comment (JmPotato) [inline tools/pd-simulator/simulator/client.go:282]: I prefer this style to avoid unnecessary if check.

```suggestion
	if err != nil {
		return err
	}
	if res.GetHeader().GetError() != nil {
		return errors.Errorf("bootstrap failed: %s", resp.Header.GetError().String())
	}
```

File: tools/pd-simulator/simulator/client.go [hunk 12]
```diff
@@ -242,8 +245,11 @@ func (c *client) AllocID(ctx context.Context) (uint64, error) {
 		Header: c.requestHeader(),
 	})
 	cancel()
-	if err != nil {
-		return 0, err
+	if err != nil || resp.GetHeader().GetError() != nil {
```
Comment (JmPotato) [inline tools/pd-simulator/simulator/client.go:248]: Ditto.

File: server/api/label_test.go [hunk 13]
```diff
@@ -289,8 +289,9 @@ func (suite *strictlyLabelsStoreTestSuite) TestStoreMatch() {
 		})
 		if testCase.valid {
 			suite.NoError(err)
+			suite.Nil(resp.Header.GetError())
```
Comment (JmPotato) [inline server/api/label_test.go:292]: Please replace all `.Header`s with `.GetHeader()`.

File: client/base_client.go [hunk 14]
```diff
@@ -329,9 +329,14 @@ func (c *baseClient) getMembers(ctx context.Context, url string, timeout time.Du
 		return nil, err
 	}
 	members, err := pdpb.NewPDClient(cc).GetMembers(ctx, &pdpb.GetMembersRequest{})
-	if err != nil {
-		attachErr := errors.Errorf("error:%s target:%s status:%s", err, cc.Target(), cc.GetState().String())
-		return nil, errs.ErrClientGetMember.Wrap(attachErr).GenWithStackByCause()
+	if err != nil || members != nil && members.Header.GetError() != nil {
```
[the remark this answers is not in the document]
Comment (HuSharp) [reply client/base_client.go:332]: > You can always use GetXXX() to avoid panic when using the generated proto buff struct.
When I entered the `GetHeader` function, I found that it just returned nil when resp was nil, which would cause fail if I continued to call `members.Header.GetError()` (before adding the judgment on members here, the git test had indicated an error: `point is nil`)

File: client/client.go [hunk 15]
```diff
@@ -585,10 +586,9 @@ func (c *client) GetAllMembers(ctx context.Context) ([]*pdpb.Member, error) {
 	ctx = grpcutil.BuildForwardContext(ctx, c.GetLeaderAddr())
 	resp, err := c.getClient().GetMembers(ctx, req)
 	cancel()
-	if err != nil {
-		cmdFailDurationGetAllMembers.Observe(time.Since(start).Seconds())
-		c.ScheduleCheckLeader()
-		return nil, errors.WithStack(err)
+
```
Comment (JmPotato) [inline client/client.go:589]:
```suggestion
```

**Review by JmPotato — COMMENTED** [review]
The rest LGTM.

**Comment by JmPotato** [comment]
@rleungx PTAL, thx!

**Review by rleungx — APPROVED** [review]

**Review by JmPotato — APPROVED** [review]
